packages/fresh/lib/src/fresh.dart

```diff
@@ -89,3 +91,122 @@ class InMemoryTokenStorage<T extends Token> implements TokenStorage<T> {
     _token = token;
   }
 }
+
+/// An interface that must be implemented to create an interceptor
+/// that transparently updates / caches tokens.
+abstract class FreshBase<T> {
+  /// {@template removeToken}
+  /// Sets the internal [token] to the provided [token]
+  /// and updates the `AuthenticationStatus` accordingly.
+  ///
+  /// If the provided token is null, the `AuthenticationStatus` will be updated
+  /// to `unauthenticated` and
+  /// and the token will be removed from storage, otherwise it
+  /// will be updated to `authenticated`
+  /// and save to storage.
+  ///
+  /// This method should be called after making a successful token request
+  /// from the custom `RefreshInterceptor` implementation.
+  ///
+  ///
+  /// If you want to remove the token, set as null and update
+  ///`AuthenticationStatus` to `unauthenticated`
+  /// you can use `removeToken()` instead of `setToken(null)`.
+  ///
+  ///
+  ///Using `removeToken()` or `setToken(null)` the behavior will be the same,
+  ///but using `removeToken()` is more clearer.
+  /// {@endtemplate}
+  Future<void> setToken(T token);
+
+  /// Removes the internal [token] from the storage, sets the internal [token]
+  ///  as null and updates the `AuthenticationStatus` to `unauthenticated`.
+  /// This method should be called when you want to log off the user.
+  Future<void> removeToken();
+
+  /// Returns a `Stream<AuthenticationState>` which is updated internally based
+  /// on if a valid token exists in [TokenStorage].
+  Stream<AuthenticationStatus> get authenticationStatus;
+
+  /// Returns a `Stream<T>` which is updated internally based
+  /// on if a valid token exists in [TokenStorage].
+  Stream<T> get currentToken;
+}
+
+/// {@template controller}
+///A token controller for handles update / caching tokens transparently.
+/// {@endtemplate}
+class FreshController<T> implements FreshBase<T> {
+  /// {@macro controller}
+  FreshController({@required TokenStorage<T> tokenStorage})
+      : _tokenStorage = tokenStorage {
+    _tokenStorage.read().then(updateStatus);
+  }
+
+  final StreamController<AuthenticationStatus> _controller =
+      StreamController<AuthenticationStatus>.broadcast()
+        ..add(AuthenticationStatus.initial);
+
+  final StreamController<T> _tokenController = StreamController<T>.broadcast();
```

Comment:
we should provide a `close` in order to close these internal `StreamControllers`

Reply:
I really forgot to provide this method.
I implemented the Sink interface to provide the closing method.In this way, it is also possible to add the token with the "add" method and  users will be advised to close this by the close_sinks rule.

Let me know what you think.